### ISSUE TITLE
refactor: concept-layer schema migration — promote CompiledClaim to a table, add concept_cluster + claim_concept (foundation for #465)

### ISSUE THREAD

**Issue by manavgup** [manavgup/wikimind#466]
> **Revised after architectural review.** See "Revision history" at the bottom for what changed and why.

## Context

#465 specified concept-level confidence scoring. This issue is the **storage substrate** it needs, while we're still at v0.1.0 and can make breaking changes without compatibility shims.

Sister issues: #422 (article-level confidence — merged), #465 (concept-level scoring — design), #444 (self-model — concepts grow toward this), #423 (typed edges — concepts will hang relationships off these).

The architecture is **Design A** from the #465 discussion (implicit concepts modeled as clusters), revised after architectural review to fix three structural problems with the original draft:

1. **Cluster on extracted subjects, not full claim text.** "Redis is fast" and "Redis is slow" must surface as opposing predicates of the same concept, not unrelated clusters.
2. **Two-stage pipeline.** Online ingest does cheap *advisory* labeling; an offline reconciler does ground-truth clustering. This avoids order-dependent, irreversible decisions at ingest time.
3. **Versioned embeddings + candidate cluster status.** Anticipates embedding-model upgrades and singleton-noise without an emergency migration later.

## Current state (post #422)

- `Article` has `confidence_score`, `last_reinforced_at`.
- `CompiledClaim` is a Pydantic model only (`models.py:434`); claims persist as a JSON `key_claims` array on the article row. No FK, no per-claim queries.
- `Backlink` is typed (#423).
- Embedding infrastructure: needs verification before this issue starts. If absent, this issue includes adding one.

## Target schema

```sql
-- Embedding type column shorthand: VECTOR(N)
-- Postgres: pgvector extension (`vector(N)`)
-- SQLite: BLOB stored as raw float32 array, queried via sqlite-vec virtual table

compiled_claim (
    id                   UUID PRIMARY KEY,
    article_id           UUID NOT NULL REFERENCES article(id) ON DELETE CASCADE,
    text                 TEXT NOT NULL,
    subjects             JSON NOT NULL DEFAULT '[]', -- list[str], LLM-extracted canonical subject names
    predicate            TEXT,                       -- LLM-extracted predicate (nullable initially)
    confidence_level     TEXT NOT NULL,              -- existing categorical enum
    confidence_score     REAL NOT NULL DEFAULT 0.5,  -- numeric, reused from #422
    source_ids           JSON NOT NULL DEFAULT '[]', -- list of source UUIDs supporting this claim
    last_reinforced_at   TIMESTAMPTZ NOT NULL,
    quote                TEXT,
    embedding            VECTOR(N),                  -- nullable until embedding pipeline runs
    embedding_version    TEXT,                       -- e.g. "bge-small-1.5" — invalidates on backend change
    cluster_assignment_reconciled BOOLEAN NOT NULL DEFAULT FALSE,
    created_at           TIMESTAMPTZ NOT NULL DEFAULT now(),
    updated_at           TIMESTAMPTZ NOT NULL DEFAULT now()
);
CREATE INDEX ix_compiled_claim_article_id ON compiled_claim(article_id);
CREATE INDEX ix_compiled_claim_last_reinforced_at ON compiled_claim(last_reinforced_at);
CREATE INDEX ix_compiled_claim_embedding_version ON compiled_claim(embedding_version);

concept_cluster (
    id                  UUID PRIMARY KEY,
    canonical_text      TEXT NOT NULL,           -- canonical subject name (LLM-suggested or member-derived)
    centroid_embedding  VECTOR(N) NOT NULL,
    embedding_version   TEXT NOT NULL,           -- centroid is only valid for this version
    member_count        INTEGER NOT NULL DEFAULT 0,
    status              TEXT NOT NULL DEFAULT 'candidate',  -- candidate|active|archived|superseded|rejected
    superseded_by       UUID REFERENCES concept_cluster(id),
    last_reinforced_at  TIMESTAMPTZ NOT NULL,
    last_reconciled_at  TIMESTAMPTZ,             -- when offline clusterer last touched this row
    created_at          TIMESTAMPTZ NOT NULL DEFAULT now(),
    updated_at          TIMESTAMPTZ NOT NULL DEFAULT now()
);
CREATE INDEX ix_concept_cluster_status ON concept_cluster(status);
CREATE INDEX ix_concept_cluster_last_reinforced_at ON concept_cluster(last_reinforced_at);
CREATE INDEX ix_concept_cluster_embedding_version ON concept_cluster(embedding_version);

claim_concept (
    claim_id    UUID NOT NULL REFERENCES compiled_claim(id) ON DELETE CASCADE,
    concept_id  UUID NOT NULL REFERENCES concept_cluster(id),
    role        TEXT NOT NULL,                   -- 'subject' | 'mentioned'
    advisory    BOOLEAN NOT NULL DEFAULT TRUE,   -- TRUE until offline reconciler confirms
    created_at  TIMESTAMPTZ NOT NULL DEFAULT now(),
    PRIMARY KEY (claim_id, concept_id, role)
);
CREATE INDEX ix_claim_concept_concept_id ON claim_concept(concept_id, role);

-- Drop from article:
ALTER TABLE article DROP COLUMN key_claims;
```

`N` (embedding dimension) is set once by the embedding model choice. Likely 384 (BGE-small / all-MiniLM-L6-v2) for local; 1536 (OpenAI ada-002) if remote. Pick at issue start, encode as `EMBEDDING_DIM` setting.

### Status semantics

- `candidate` — singleton or unconfirmed; not yet promoted. Hidden from default views.
- `active` — promoted (member_count ≥ 2 AND reconciled by offline job at least once).
- `archived` — no reinforcement for >N months; recoverable.
- `superseded` — merged into another cluster; `superseded_by` redirects.
- `rejected` — flagged as a bad cluster; kept as negative training data.

### Online vs offline assignment

- `claim_concept.advisory = TRUE` is the default at ingest time — the online clusterer's best guess.
- The offline reconciler (see "Code surface") sets `advisory = FALSE`, updates `cluster_assignment_reconciled = TRUE`, and stamps `concept_cluster.last_reconciled_at`.
- Q&A and graph endpoints can choose to filter by `advisory = FALSE` when answer quality matters more than completeness.

## Migration sequence

Three Alembic migrations.

### `0008_embedding_infrastructure.py`

- Postgres: `op.execute("CREATE EXTENSION IF NOT EXISTS vector")`
- SQLite: no-op (sqlite-vec loaded at runtime via `conn.load_extension()`)
- Add a SQLAlchemy `TypeDecorator` `Vector(dim)` in `src/wikimind/db_types.py` that maps to `vector(N)` on Postgres and stores raw bytes on SQLite. Hide the dialect branch behind it so models stay clean.

### `0009_compiled_claim_table.py`

Promote `CompiledClaim` from JSON-on-article to a proper table.

- Create `compiled_claim` table per schema above (including `subjects`, `embedding_version`, `cluster_assignment_reconciled`).
- **No backfill.** v0.1.0; recompile is the documented upgrade path. Migration drops `article.key_claims`.
- Update `CompiledClaim` in `models.py` from a `BaseModel` to an SQLModel table.
- Update `Compiler.save_article()` to insert claim rows after the article row commits.

### `0010_concept_cluster_and_join.py`

- Create `concept_cluster` and `claim_concept` tables per schema above (including `candidate` default status, `embedding_version`, `last_reconciled_at`, `advisory` flag).
- No backfill.

## Code surface

| Path | Purpose |
|---|---|
| `src/wikimind/db_types.py` (new) | `Vector(dim)` TypeDecorator |
| `src/wikimind/engine/embeddings.py` (new) | One-call `embed(text) -> list[float]`; backend BGE-small (default) or OpenAI. Records `embedding_version` constant. |
| `src/wikimind/engine/concept_store.py` (new) | Repository: `find_or_create_cluster(claim_subjects, claim_embedding) -> ConceptCluster`. **Online (advisory) labeling only.** ANN top-K → auto-join >0.85 → LLM tiebreaker [0.70, 0.85] → new `candidate` cluster. |
| `src/wikimind/engine/concept_reconciler.py` (new) | **Offline reconciler.** Batch HDBSCAN over all `compiled_claim.embedding` rows of the current `embedding_version`. Reconciles cluster membership, promotes `candidate → active` (member_count ≥ 2), proposes splits/merges. Writes audit log entries. Run via `wikimind reconcile-concepts` CLI command and a future scheduled job. |
| `src/wikimind/engine/compiler.py` | Compiler prompt extension: structured-output schema gains `subjects: list[str]` and (optional) `predicate: str` per claim. After saving claims: embed each, call `concept_store.find_or_create_cluster` per subject, insert `claim_concept` rows with `advisory=TRUE`. |
| `src/wikimind/models.py` | Promote `CompiledClaim` to SQLModel table; add `ConceptCluster`, `ClaimConcept` SQLModels. Drop `Article.key_claims`. |
| `src/wikimind/api/routes/wiki.py` | New `GET /wiki/concepts/{id}` (cluster + member claims), `GET /wiki/articles/{id}/concepts`. Both accept `?include_advisory=true` (default false on Q&A-quality paths). |
| `src/wikimind/api/routes/admin.py` (or extend existing) | `POST /admin/reconcile-concepts` — triggers the offline reconciler. Same code path as the CLI command. Returns a job summary. |

## Compiler prompt change

Extend the existing structured-output schema for compilation. Per claim:

```
{
  "claim": "Redis is single-threaded",
  "subjects": ["Redis", "single-threaded execution"],
  "predicate": "is",
  "confidence": "SOURCED",
  "quote": "...",
  "source_ids": ["src_abc", "src_def"]
}
```

Subjects are **canonical names the LLM picks** — they don't need to match anything yet. The reconciler is responsible for grouping "Redis" and "redis (in-memory K/V store)" into one cluster downstream.

## Backfill strategy

**None.** v0.1.0 means no production data covenant. First run after merge requires a recompile of any existing wiki articles. Document in:
- The migration file's docstring
- The PR body
- README under "Upgrading" (add the section if absent)

A `make recompile-all` target ships in this PR — without it, anyone pulling main has a broken wiki until they figure out how to recompile.

## Configuration additions

```python
class ConceptConfig(BaseModel):
    embedding_backend: Literal["bge-small", "openai"] = "bge-small"
    embedding_dim: int = 384
    embedding_version: str = "bge-small-1.5"  # bumped on backend change
    auto_join_threshold: float = 0.85
    tiebreaker_threshold: float = 0.70
    cluster_promote_min_members: int = 2          # candidate → active
    cluster_canonical_refresh_min_members: int = 5
    cluster_size_split_threshold: int = 100       # auto-flag for split when exceeded
    reconciler_min_interval_minutes: int = 60     # rate-limit /admin/reconcile-concepts
```

Update `.env.example`.

## Observability — required, not deferred

These metrics ship as part of acceptance, exposed via a `/admin/concepts/metrics` endpoint and structured logs:

- Cluster count, broken down by `status`
- Member-count distribution histogram (singleton rate is a key signal)
- Tiebreaker LLM invocation rate (% of claims hitting the [0.70, 0.85] window)
- Tiebreaker verdict distribution (same vs different)
- Per-claim ingest decision latency (P50, P99)
- Orphan claim rate (% of `compiled_claim` rows with no `claim_concept` row — must be 0 in steady state)
- Reconciler last-run timestamp, runtime, claims reconciled, splits proposed, merges proposed
- Count of claims with stale `embedding_version` (≠ current `ConceptConfig.embedding_version`)

## Failure modes addressed in this design

- **Embedding model upgrades** — `embedding_version` per row + reconciler can detect mismatch and re-embed lazily. Cluster IDs survive across re-embeddings because reconciler reuses canonical_text as anchor.
- **Order dependence at ingest** — online assignment is `advisory`; offline reconciler is the source of truth.
- **Cold-start singleton noise** — `candidate` status hides single-member clusters from default views until promoted.
- **Skewed corpora** — `cluster_size_split_threshold` flags oversized clusters for the reconciler to split.
- **Tiebreaker LLM non-determinism** — every tiebreaker call logs verdict + similarity; agreement rate is a metric.
- **Multi-user leakage (when #364 lands)** — clusters must be user-scoped. Add `user_id` to `concept_cluster` and `compiled_claim` if multi-user is already in the model elsewhere; otherwise leave a TODO note for the multi-user issue.

## Testing approach

- **Migration tests**: forward-only round-trip on SQLite and Postgres in CI. CI must run both — vector type behavior diverges.
- **`embeddings.py` unit tests**: mock backend, assert dimension, determinism, version stamp.
- **`concept_store.find_or_create_cluster` table-driven tests**: high-similarity → join, mid-similarity → tiebreaker called, low → new candidate. Mock LLM tiebreaker.
- **`concept_reconciler` tests**: synthetic embedding sets where HDBSCAN should produce K clusters; assert produced, advisory→reconciled flag flip, candidate→active promotion at member_count threshold.
- **Compiler integration**: ingest two paraphrased sources of the same fact, verify (a) both produce claims with overlapping subjects, (b) online clustering puts them in one cluster (advisory), (c) running the reconciler confirms or refines.
- **Snapshot test** for `GET /wiki/concepts/{id}` and `GET /wiki/articles/{id}/concepts`.
- **Observability test**: `/admin/concepts/metrics` returns the documented shape.

## Out of scope (deferred to follow-ups)

- **Concept promotion (cluster → first-class `Concept` row with description, manual canonical_name override).** Schema is forward-compatible.
- **Concept hierarchy** (`parent_concept_id`).
- **User feedback as must-link / cannot-link constraints to the reconciler.**
- **UI surfaces** for concepts.
- **Reinforcement math wiring** (the cluster row exists; recompute on reinforcement is #465's work).
- **Source-quality-weighted confidence** (still using #422's 0.7 stub).
- **Subject/predicate fully utilized for contradiction detection** — predicate column exists, but using it is #465.

## Open decisions to make at PR time

1. **Embedding backend default.** BGE-small (local, free) vs OpenAI (better quality). Recommendation: BGE-small.
2. **Embedding dimension.** Tied to (1). 384 vs 1536.
3. **Tiebreaker LLM.** Reuse compiler client. Recommendation: yes.
4. **Multi-user scoping.** Check whether `user_id` is already on `Article` and friends. If yes, add it to new tables now; if no, leave as TODO for #364.

## Acceptance criteria

- [ ] Three migrations land cleanly on fresh SQLite and fresh Postgres
- [ ] `make verify` passes including migration round-trip tests
- [ ] Compiler produces persisted `compiled_claim` rows with `subjects`, `embedding`, `embedding_version` on a fresh ingest
- [ ] Two paraphrased sources of the same fact produce claims that share at least one cluster (advisory)
- [ ] `wikimind reconcile-concepts` CLI runs end-to-end on a populated DB and flips `advisory=FALSE` for reconciled rows
- [ ] `candidate` clusters with member_count ≥ 2 are promoted to `active` by the reconciler
- [ ] `GET /wiki/concepts/{id}` and `GET /wiki/articles/{id}/concepts` work, with `?include_advisory` filter
- [ ] `GET /admin/concepts/metrics` returns the documented metric set
- [ ] README / upgrade notes call out the v0.1.0 → post-concept recompile requirement
- [ ] OpenAPI regenerated

## Complexity

Medium-high. The migrations are mechanical; the real work is the embedding service, the online clusterer, and the offline reconciler. Two-stage architecture adds ~30% effort vs the original online-only draft, but heads off a foreseeable rewrite at v0.2.0.

Estimate: 5–7 days for a focused implementer, with threshold tuning extending into a follow-up tuning issue.

## Revision history

**v2 (post-architectural-review)** — three structural changes from the original draft:

1. **Cluster on extracted subjects, not full claim text.** Original schema would have produced "claim clusters" mis-named as concept clusters; "Redis is fast" and "Redis is slow" would not have surfaced as opposing predicates of the same concept. Subject extraction (`subjects: list[str]`) is added to the compiler prompt and persisted on `compiled_claim`. The reconciler clusters on these subjects.
2. **Two-stage clustering.** Online ingest now does *advisory* labeling (`claim_concept.advisory = TRUE`); a new `concept_reconciler` (CLI + admin endpoint) does ground-truth clustering offline. Avoids order-dependent, irreversible decisions and supports re-clustering on embedding-model upgrades or new constraints.
3. **`embedding_version` columns + `candidate` cluster status + observability acceptance criteria.** Anticipates embedding upgrades, hides singleton noise from default views, and surfaces metrics needed to know whether the layer is silently broken.

**v1 (original)** — online-only clustering on full claim text, no `embedding_version`, no `candidate` status, observability deferred. Superseded by v2.
